File: scripts/render-player-motion.mjs
// CPU reference capture of actual runtime skin deformation. No WebGL/PBR claim.
// node scripts/render-player-motion.mjs [/tmp/nesi-player-motion]
// python3 scripts/render-player-motion.py /tmp/nesi-player-motion /tmp/nesi-raster
import fs from 'node:fs';
import vm from 'node:vm';
import path from 'node:path';
import assert from 'node:assert/strict';
import * as THREE from 'three';
import { LabPlayerAnimator } from '../src/game/LabPlayerAnimator.js';
import { LabHeldDevice } from '../src/game/LabHeldDevice.js';
const output = path.resolve(process.argv[2] || '/tmp/nesi-player-motion');
fs.mkdirSync(output, { recursive: true });
const scope = { console, TextDecoder, module: { exports: {} } };
vm.runInNewContext(fs.readFileSync(new URL('../node_modules/three/examples/jsm/libs/draco/gltf/draco_decoder.js', import.meta.url), 'utf8'), scope);
const draco = await scope.DracoDecoderModule();
async function load(file, textureName) {
  const bytes = fs.readFileSync(file), length = bytes.readUInt32LE(12);
  const doc = JSON.parse(bytes.subarray(20, 20 + length)), bin = bytes.subarray(28 + length);
  const components = { SCALAR: 1, VEC2: 2, VEC3: 3, VEC4: 4, MAT4: 16 };
  const types = { 5120: [Int8Array, 1, 'getInt8'], 5121: [Uint8Array, 1, 'getUint8'],
    5122: [Int16Array, 2, 'getInt16'], 5123: [Uint16Array, 2, 'getUint16'],
    5125: [Uint32Array, 4, 'getUint32'], 5126: [Float32Array, 4, 'getFloat32'] };
  function accessor(id) {
    const acc = doc.accessors[id], view = doc.bufferViews[acc.bufferView];
    const [Type, size, getter] = types[acc.componentType], width = components[acc.type];
    const data = new DataView(bin.buffer, bin.byteOffset, bin.byteLength), array = new Type(acc.count * width);
    const start = (view.byteOffset || 0) + (acc.byteOffset || 0), stride = view.byteStride || size * width;
    for (let i = 0; i < acc.count; i++) for (let j = 0; j < width; j++) array[i * width + j] = data[getter](start + i * stride + j * size, true);
    return new THREE.BufferAttribute(array, width, !!acc.normalized);
  }
  function mesh(primitive) {
    const geometry = new THREE.BufferGeometry(), ext = primitive.extensions?.KHR_draco_mesh_compression;
    const semantics = [['POSITION','position',3],['NORMAL','normal',3],['TEXCOORD_0','uv',2]];
    if (ext) {
      const view = doc.bufferViews[ext.bufferView], buffer = new draco.DecoderBuffer();
      buffer.Init(new Int8Array(bin.subarray(view.byteOffset || 0, (view.byteOffset || 0) + view.byteLength)), view.byteLength);
      const decoder = new draco.Decoder(), decoded = new draco.Mesh();
      assert.ok(decoder.DecodeBufferToMesh(buffer, decoded).ok(), `Decode failed: ${file}`);
      for (const [semantic, name, size] of semantics) {
        if (!(semantic in ext.attributes)) continue;
        const attribute = decoder.GetAttributeByUniqueId(decoded, ext.attributes[semantic]);
        const data = new draco.DracoFloat32Array(); decoder.GetAttributeFloatForAllPoints(decoded, attribute, data);
        const array = new Float32Array(decoded.num_points() * size);
        for (let i = 0; i < array.length; i++) array[i] = data.GetValue(i);
        geometry.setAttribute(name, new THREE.BufferAttribute(array, size)); draco.destroy(data);
      }
      const face = new draco.DracoInt32Array(), indices = new Uint32Array(decoded.num_faces() * 3);
      for (let i = 0; i < decoded.num_faces(); i++) { decoder.GetFaceFromMesh(decoded, i, face); for (let k = 0; k < 3; k++) indices[i * 3 + k] = face.GetValue(k); }
      geometry.setIndex(new THREE.BufferAttribute(indices, 1));
      draco.destroy(face); draco.destroy(decoded); draco.destroy(decoder); draco.destroy(buffer);
    } else {
      for (const [semantic, name] of semantics) if (semantic in primitive.attributes) geometry.setAttribute(name, accessor(primitive.attributes[semantic]));
      if (primitive.indices !== undefined) geometry.setIndex(accessor(primitive.indices));
    }
    const material = doc.materials[primitive.material];
    const texture = doc.textures[material.pbrMetallicRoughness.baseColorTexture.index];
    const source = texture.extensions?.EXT_texture_webp?.source ?? texture.source;
    const image = doc.images[source], imageView = doc.bufferViews[image.bufferView];
    fs.writeFileSync(`${output}/${textureName}.webp`, bin.subarray(imageView.byteOffset || 0, (imageView.byteOffset || 0) + imageView.byteLength));
    return new THREE.Mesh(geometry, new THREE.MeshStandardMaterial({ side: THREE.DoubleSide }));
  }
  const nodes = doc.nodes.map(node => {
    const object = new THREE.Group();
    if (node.mesh !== undefined) for (const primitive of doc.meshes[node.mesh].primitives) object.add(mesh(primitive));
    if (node.rotation) object.quaternion.fromArray(node.rotation);
    if (node.translation) object.position.fromArray(node.translation);
    if (node.scale) object.scale.fromArray(node.scale);
    if (node.matrix) new THREE.Matrix4().fromArray(node.matrix).decompose(object.position, object.quaternion, object.scale);
    return object;
  });
  doc.nodes.forEach((node, i) => { for (const child of node.children || []) nodes[i].add(nodes[child]); });
  const group = new THREE.Group(); for (const id of doc.scenes[doc.scene || 0].nodes) group.add(nodes[id]); return group;
}

function normalize(source, size, height = false) {
  const root = new THREE.Group();
  let box = new THREE.Box3().setFromObject(source), dimensions = box.getSize(new THREE.Vector3());
  source.scale.multiplyScalar(size / (height ? dimensions.y : Math.max(...dimensions.toArray())));
  box = new THREE.Box3().setFromObject(source);
  const center = box.getCenter(new THREE.Vector3());
  source.position.add(new THREE.Vector3(-center.x, -box.min.y, -center.z));
  root.add(source); return root;
}
const root = new THREE.Group(), carrier = new THREE.Group(); root.add(carrier);
const visual = normalize(await load(new URL('../public/models/runtime/model-01-player.glb', import.meta.url), 'player'), 2.4, true);
root.add(visual);
const animator = new LabPlayerAnimator({ visual, carrier });
const gun = normalize(await load(new URL('../public/models/runtime/model-11-portal-gun.glb', import.meta.url), 'gun'), .7);
const held = new LabHeldDevice({ model: gun, bones: animator.bones, playerRoot: root, size: .7 });
const streams = [];
function geometry(mesh, key) {
  const g = mesh.geometry;
  fs.writeFileSync(`${output}/${key}-uv.bin`, Buffer.from(g.attributes.uv.array.buffer));
  const indices = Uint32Array.from(g.index.array);
  fs.writeFileSync(`${output}/${key}-indices.bin`, Buffer.from(indices.buffer));
  streams.push({ key, mesh });
}
geometry(animator.rig.mesh, 'player');
gun.traverse(object => { if (object.isMesh) geometry(object, 'gun'); });
const frames = [];
const point = new THREE.Vector3();
function capture(name, index, time) {
  root.updateMatrixWorld(true); animator.rig.skeleton.update();
  const meshes = [];
  for (const { key, mesh } of streams) {
    const g = mesh.geometry, positions = new Float32Array(g.attributes.position.count * 3);
    for (let i = 0; i < g.attributes.position.count; i++) {
      mesh.getVertexPosition(i, point).applyMatrix4(mesh.matrixWorld).sub(root.position);
      point.toArray(positions, i * 3);
    }
    const normalGeometry = new THREE.BufferGeometry();
    normalGeometry.setAttribute('position', new THREE.BufferAttribute(positions, 3)); normalGeometry.setIndex(g.index);
    normalGeometry.computeVertexNormals();
    const prefix = `${name}-${String(index).padStart(3, '0')}-${key}`;
    fs.writeFileSync(`${output}/${prefix}-position.bin`, Buffer.from(positions.buffer));
    fs.writeFileSync(`${output}/${prefix}-normal.bin`, Buffer.from(normalGeometry.attributes.normal.array.buffer));
    normalGeometry.dispose(); meshes.push({ key, prefix });
  }
  frames.push({ name, index, time, y: root.position.y, meshes, cadence: animator.cadenceHz,
    phase: animator.diagnostics.airborne.phase, contacts: animator.diagnostics.groundContact });
}
function step({ speed = 0, carrying = false, grounded = true, vy = 0, y = 0 } = {}) {
  const dt = 1 / 60; root.position.z += speed * dt; root.position.y = y;
  const targets = carrying ? {
    left: root.position.clone().add(new THREE.Vector3(-.20, 1.045, .52)),
    right: root.position.clone().add(new THREE.Vector3(.20, 1.045, .52)),
  } : null;
  animator.update({ dt, speed, velocity: new THREE.Vector3(0, vy, speed), grounded, carrying,
    carryGripTargets: targets, sampleGround: () => ({ height: 0, normal: new THREE.Vector3(0, 1, 0) }) });
  held.update({ dt, carrying });
}
for (const [name, speed, carrying] of [['idle', 0, false], ['walk', 3.3, false], ['run', 5, false], ['carry', 2.9, true]]) {
  animator.reset(); held.reset(); root.position.set(0, 0, 0);
  for (let i = 0; i < 120; i++) step({ speed, carrying });
  for (let i = 0; i < 90; i++) { step({ speed, carrying }); if (i % 3 === 0) capture(name, i / 3, i / 60); }
  console.log(`Captured ${name}: cadence ${animator.cadenceHz.toFixed(3)} cycles/s`);
}
animator.reset(); held.reset(); root.position.set(0, 0, 0);
for (let i = 0; i < 90; i++) step({ speed: 3.3 });
animator.triggerJump();
for (let i = 0; i < 66; i++) {
  const age = i / 60 - .05, air = age > 0 && age < 2 * 7.8 / 19.5;
  step({ speed: 3.3, grounded: !air, vy: air ? 7.8 - 19.5 * age : 0,
    y: air ? 7.8 * age - 19.5 * age * age / 2 : 0 });
  if (i % 2 === 0) capture('jump', i / 2, i / 60);
}
fs.writeFileSync(`${output}/frames.json`, JSON.stringify({ method: 'Actual runtime mesh, production LabPlayerAnimator and foot contact, CPU skinning; orthographic base texture capture; no browser/PBR claim', frames }, null, 2));
console.log(`Exported ${frames.length} mesh frames to ${output}`);

File: scripts/render-player-motion.py
"""Rasterize production mesh poses exported by render-player-motion.mjs.

python3 scripts/render-player-motion.py /tmp/nesi-player-motion /tmp/nesi-raster
The raster executable is compiled from optimize-runtime-models-raster.cpp.
Outputs are reference captures, never described as actual browser rendering.
"""
import json
import math
import pathlib
import struct
import subprocess
import sys

import numpy as np
from PIL import Image, ImageDraw

source = pathlib.Path(sys.argv[1])
raster = sys.argv[2]
destination = pathlib.Path(__file__).resolve().parents[1] / 'qa'
destination.mkdir(exist_ok=True)
manifest = json.loads((source / 'frames.json').read_text())
width, height = 300, 360
textures = {key: np.asarray(Image.open(source / f'{key}.webp').convert('RGBA').resize((512, 512))) for key in ['player', 'gun']}
atlas = np.concatenate(list(textures.values()), axis=1)
attributes = {}
for number, key in enumerate(textures):
    uv = np.fromfile(source / f'{key}-uv.bin', np.float32).reshape(-1, 2)
    uv[:, 0] = (uv[:, 0] * 511 + number * 512) / 1023
    attributes[key] = (uv, np.fromfile(source / f'{key}-indices.bin', np.uint32))

def render(frame, yaw=.68, world_arc=False):
    pitch = .10
    camera = np.array([math.sin(yaw) * math.cos(pitch), math.sin(pitch), math.cos(yaw) * math.cos(pitch)])
    right = np.cross([0, 1, 0], camera); right /= np.linalg.norm(right)
    up = np.cross(camera, right)
    light = np.array([-.4, .8, .5]); light /= np.linalg.norm(light)
    vertices, faces, offset = [], [], 0
    for mesh in frame['meshes']:
        p = np.fromfile(source / f"{mesh['prefix']}-position.bin", np.float32).reshape(-1, 3)
        n = np.fromfile(source / f"{mesh['prefix']}-normal.bin", np.float32).reshape(-1, 3)
        local = p - [0, 2.05 if world_arc else 1.20, 0]
        if world_arc:
            local[:, 1] += frame['y']
        scale = 76 if world_arc else 125
        vert = np.empty((len(p), 6), dtype=np.float32)
        vert[:, 0] = width * .5 + local @ right * scale
        vert[:, 1] = height * .52 - local @ up * scale
        vert[:, 2] = local @ camera
        vert[:, 3] = .65 + .35 * np.maximum(n @ light, 0)
        uv, indices = attributes[mesh['key']]
        vert[:, 4:] = uv
        vertices.append(vert); faces.append(indices + offset); offset += len(p)
    vertices, faces = np.concatenate(vertices), np.concatenate(faces)
    binary, output = source / 'raster-input.bin', source / 'raster-output.rgba'
    with binary.open('wb') as f:
        f.write(struct.pack('<6I', width, height, len(vertices), len(faces), atlas.shape[1], atlas.shape[0]))
        f.write(vertices.tobytes()); f.write(faces.tobytes()); f.write(atlas.tobytes())
    subprocess.run([raster, str(binary), str(output)], check=True)
    pixels = Image.fromarray(np.fromfile(output, np.uint8).reshape(height, width, 4), 'RGBA')
    image = Image.new('RGB', (width, height), '#e9eef2')
    draw = ImageDraw.Draw(image)
    draw.line([(12, 340), (width-12, 340)], fill='#c6d2da', width=1)
    image.paste(pixels, (0, 0), pixels)
    draw.text((12, 9), f"{frame['name']}   {frame['time']:.2f}s", fill='#253c4e')
    if frame['name'] == 'jump': draw.text((12, 26), frame['phase'] + f" / y={frame['y']:.2f}m", fill='#405b6b')
    return image

names = ['idle', 'walk', 'run', 'jump', 'carry']
sheet = Image.new('RGB', (6*width, 5*(height+28)+52), '#ffffff')
draw = ImageDraw.Draw(sheet)
draw.text((16, 12), 'NESI v6 | Actual runtime mesh + production animation | CPU reference capture, not browser/PBR', fill='#213947')
draw.text((16, 29), 'Frames follow the player root. Jump height is written in metres. Carry checks the wrist pose; companion is omitted.', fill='#516470')
stats = {}
reference = np.fromfile(source / 'idle-000-player-position.bin', np.float32).reshape(-1, 3)
boot_masks = {'left': (reference[:, 1] < .30) & (reference[:, 0] < 0),
              'right': (reference[:, 1] < .30) & (reference[:, 0] > 0)}
for row, name in enumerate(names):
    frames = [frame for frame in manifest['frames'] if frame['name'] == name]
    select = np.linspace(0, len(frames)-1, 6).astype(int)
    for column, index in enumerate(select):
        sheet.paste(render(frames[index]), (column*width, 52+row*(height+28)))
    label = f"{name.upper()} | {frames[-1]['cadence']:.2f} cycles/s" if name != 'jump' else 'JUMP | takeoff - tuck - descent - landing'
    draw.text((16, 52+row*(height+28)+height+7), label, fill='#253c4e')
    if name in ('run', 'walk', 'jump'):
        sequence = [render(frame) for frame in frames]
        sequence[0].save(destination / f'player-motion-v6-{name}.webp', save_all=True, append_images=sequence[1:], duration=33 if name == 'jump' else 50, loop=0, quality=82, method=4)
        if name == 'jump':
            sequence = [render(frame, world_arc=True) for frame in frames]
            sequence[0].save(destination / 'player-motion-v6-jump-arc.webp', save_all=True, append_images=sequence[1:], duration=33, loop=0, quality=82, method=4)
    stats[name] = {'cadenceHz': frames[-1]['cadence'], 'frames': len(frames), 'phases': sorted(set(f['phase'] for f in frames))}
    stats[name]['bootMeshMeasurements'] = {}
    for side, mask in boot_masks.items():
        points = np.array([np.fromfile(source / f"{frame['meshes'][0]['prefix']}-position.bin", np.float32).reshape(-1, 3)[mask].mean(0) for frame in frames])
        stats[name]['bootMeshMeasurements'][side] = {
            'rootRelativeCenterRangeMetres': np.ptp(points, axis=0).round(6).tolist(),
            'maximumCenterDisplacementMetres': float(np.linalg.norm(np.diff(points, axis=0), axis=1).max()),
            'sampleIntervalSeconds': frames[1]['time'] - frames[0]['time'],
        }
    print(f'Rendered {name}', flush=True)
sheet.save(destination / 'player-motion-v6-contact-sheet.jpg', quality=92)
# A side view exposes knee direction and boot contact that a front view hides.
side = Image.new('RGB', (6*width, 2*(height+28)+32), '#ffffff')
for row, name in enumerate(['run', 'jump']):
    frames = [frame for frame in manifest['frames'] if frame['name'] == name]
    for column, index in enumerate(np.linspace(0, len(frames)-1, 6).astype(int)):
        side.paste(render(frames[index], math.pi/2), (column*width, 32+row*(height+28)))
ImageDraw.Draw(side).text((16, 10), 'Side reference | actual deformed mesh | run / jump', fill='#253c4e')
side.save(destination / 'player-motion-v6-side.jpg', quality=92)
(destination / 'player-motion-v6-report.json').write_text(json.dumps({'method': manifest['method'], 'states': stats}, indent=2)+'\n')
print(destination / 'player-motion-v6-contact-sheet.jpg')
